### DISCUSSION TITLE
How can my ILoggerProvider get a DisposeAsync() or any kind of shutdown message

### DISCUSSION THREAD

**Discussion by DavidThielen** [dotnet/aspnetcore#51357] [General]
**Blazor server**

I have just written an ILoggerProvider and the one thing it is missing is a way to shut down when the app closes.

This is not a biggie as everything is written to the log repository. But I dislike leaving things like file handles open when closing an application.

So, is there a way to be notified that the app is closing? And as a logger, be the last to be notified in case any other parts are logging while closing down?

thanks - dave

**Comment by david-acker** [comment]
Could you implement `IAsyncDisposable` on your custom ILoggerProvider, and then handle any shutdown/clean up logic in the `DisposeAsync` method implemented by that?

Since `ILoggerProvider` also implements `IDisposable`, here's some guidance on [implementing both the dispose and async dispose patterns](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#implement-both-dispose-and-async-dispose-patterns).

**Reply by DavidThielen** [reply]
I have those written. They're never called. I'm guessing the logging system never closes down its providers.

**Reply by david-acker** [reply]
Hmm, the custom `ILoggerProvider` is registered with the DI container, correct? Are there any other logging-related libraries being used?

**Reply by DavidThielen** [reply]
I'm using 4 of the standard loggers. The way mine are added is as follows:

```cs
	public static class LoggingBuilderExtensions
	{
		public static ILoggingBuilder AddFileLogger(this ILoggingBuilder builder)
		{
			builder.AddConfiguration();

			builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider,
				FileLoggerProvider>());
			builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IConfigureOptions<FileLoggerOptions>,
				FileLoggerOptionsSetup>());
			builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IOptionsChangeTokenSource<FileLoggerOptions>,
				LoggerProviderOptionsChangeTokenSource<FileLoggerOptions, FileLoggerProvider>>());
			return builder;
		}

		public static ILoggingBuilder AddAzureBlobLogger(this ILoggingBuilder builder)
		{
			builder.AddConfiguration();

			builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider,
				BlobLoggerProvider>());

			builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IConfigureOptions<BlobLoggerOptions>,
				BlobLoggerOptionsSetup>());
			builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IOptionsChangeTokenSource<BlobLoggerOptions>,
				LoggerProviderOptionsChangeTokenSource<BlobLoggerOptions, BlobLoggerProvider>>());

			return builder;
		}
	}
```
So they're singletons. Are singleton services disposed when the system shuts down? Or are they left alone? (I could probably make a good argument for either approach.)

thanks - dave

**Reply by david-acker** [reply]
In this situation, the DI container should automatically handle disposing of the `ILoggerProvider`, since its handles creating the instance on its own. To my knowledge, the only time when the container won't handle disposal is when create an instance of a service on your own and then register it (e.g. `builder.Services.AddSingleton(new ExampleService())`). [Disposal of services](https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines#disposal-of-services)

It might be worth asking this question in the [discussions](https://github.com/dotnet/runtime/discussions) for the dotnet/runtime repo to see if they have any insight on this issue since that's actually where `ILoggerProvider` and `ILoggingBuilder` are defined.

**Reply by DavidThielen** [reply]
Thank you -[ I just did](https://github.com/dotnet/runtime/issues/93562).
